[1주차] 숫자 야구게임 미션 제출합니다.

yoonjaehong26 opened the pull request:
안녕하십니까 그리디 3기 프론트앤드로 들어오게 된 윤재홍입니다.

먼저 시간을 내주셔서 코드를 읽어주시고 리뷰를 해주시는 송혜정 리뷰어님께 감사드립니다.
아직 학기가 시작된 지 얼마 되지 않았지만 앞으로 함께하면서 더 많은 얘기를 나누면 좋을 것 같습니다!

마구잡이식으로 인터넷과 책을 찾아가며 기능만 구현하면서 웹공부를 하던 저에게 이번 미션은 체계적으로 코드를 구상하고 부족한 부분을 공부할 수 있었던 좋은 기회였습니다. 

***

다음은 코드를 구상하며 중점적으로 생각했던 초기 목표들입니다.
	1.체계적인 코드구성을 위해 기능들을 최대한 함수로 구현
	2.다른사람들간의 협업을 위한 코드컨벤션 준수
	3.유지보수 및 확장성을 위해 js파일 분산화
	4.복잡하거나 신기능보다는 최대한 기본적인 코드들로 기능 구하여 사이드이펙트 최소화

***

다음은 코드를 구상하며 겪은 어려움들과 해결방안들입니다.
	1.< input>태그에서 입력데이터 프롬프트 인젝션방지
		=>초기에는 배열로 변환 후 유효성검사를 하였으나 배열로 변환과정에서 사이드이펙트가 있을 것을 예상하여
		   함수를 통한 유효성검사처리 후 배열로 변환


2.클라스,매서드에 대한 이해도 부족
		=>지금까지 c++,c#,js에서 오브젝트를 다루는 정도로 간단하게만 사용하던 클라스를, 
		프로그렘 요구사항과 같이 코드 구조로서의 개념이 와 닿지 않아 처음에는 클라스 없이 코드를 짜나가다
		막혀 올려주신 공식문서를 통해 다시 공부하고 ‘오브젝트’를 위한 기능이 아닌 ‘작은 프로그램’을 위한 기능으로 	
                이해하게 되어 코드를 작성하게 되었습니다.

3.코드 커밋 단위 구상의 어려움
		=>위에서 말씀드린 대로 최대한 함수단위로 프로그렘을 작성하고 싶어 함수를 기준으로 커밋단위를 
			구상하였으나 함수로 구현되지 않는 부분,  (feat(index.js):함수통합 밎 class구현)와 같은 코드구조 변환
			의 커밋에서 리뷰어가 이해하기 쉬운 단순한 변화가 아니게 되어 이번 과제 중 가장 아쉬운 부분이 되었습니다.


4.코드 컨벤션 밎 가독성 증가를 위한 고민
		=>원래 제가봐도 제 코드를 못알아볼정도로 기능구현만 하고 넘어갔는데 이번 과제를 기회로
		    누가봐도 이해할만한 코드를 작성하려고 노력하여 변수,함수명,코드구조를 최대한 고민해 보았으나
		    아직 다른 사람들과의 코드공유경혐이 없어 노력하고 있습니다.

***


질문드리고 싶은 점

1.  위 첫 질문과 같이 타입스크립트가 아닌 js에서 데이터 타입 핸들링하는 더 나은 방법이 있는지 궁금합니다.

2. 코드구조나 함수,변수명들이 가독성이 괜찮은지와 개선방안이 궁금합니다.

 3.코드 커밋단위 구상에 대한 개선점이 궁금합니다.

***
다음은 기능구현 gif와 dev run test 결과입니다.
부족하나 저의 PR을 읽어주셔서 감사드리고 다음주에는 더 나은 모습 보여드리기 위해 노력하겠습니다!

![1주차숫자야구게임구현영상](https://github.com/user-attachments/assets/b3cf0394-34cc-4c39-a57d-415ca282d8a6)
<img width="863" height="611" alt="스크린샷 2025-09-12 오후 12 48 59" src="https://github.com/user-attachments/assets/2b6e3d25-15ac-49e1-b06d-fae8a15bbebe" />

Commit:
docs(README) : 기능 목록 추가

Commit:
feat(index.js) : 유효한 랜덤변수 생성

Commit:
feat(index.html,index.js) : 입력데이터 리스너생성

Commit:
feat(index.js) : 입력데이터 저장

Commit:
feat(checkInput.js):배열길이검사

Commit:
feat(checkInput.js):배열중복검사

Commit:
feat(checkInput.js):배열값타당성검사

Commit:
fix(index.js,checkInput.js):import오류수정

import대상범위 오류로 인해 import대상변경

Commit:
feat(index.js):입력데이터 유효성검사

Commit:
feat(index.js):야구게임 로직구현

Commit:
feat(index.js, uiChanger.js):ui변경

Commit:
feat(index.js):함수통합 밎 class구현

Commit:
fix(iddex.js,index.html):파일import오류수정

Commit:
feat(index.js):재시작기능구현

재시작기능이 없어 재시작시 변수재설정기능을 추가

Commit:
feat(index.js,uiChanger.js):재시작 ui초기화

Commit:
refacter(index.js,checkInput.js):데이터 자리수 소프트코드

코드 확장성을 위하여 숫자야구데이터의 자리수를 3에서 변수로 변경

Commit:
refacter(index.js,uiChanger.js):불명확한 함수명 변경

가독성과 통일성이 떨어지는 함수이름 변경:changeHrml=>changeUI,initEventListeners=>uiEventListeners,handlesummit=>gameLogic

Commit:
refactor(index.js):함수명 통일성 준수

 함수명의 통일일성을 위해 함수명 변경함 gmaeLogic-유지,  handleRestart=>restartLogic

Commit:
fix(index.js):restartLogic 함수명 정정

이전 커밋에서 바꾸지 못한 함수명 정정 uiEventListeners내
handleRestart=>restartLogic

Commit:
feat(index.js)함수문법 가독성개선

함수.bind(this)문법 가독성 감소하여 화살표함수로 대체

Songhyejeong commented on src/index.js:
```diff
@@ -0,0 +1,83 @@
+  }
+
+  getPlayResultString(strike, ball) {
+    if (strike === 0 && ball === 0) {
```
> play(컴퓨터의 랜덤 값, 유저의 입력 값) 메서드를 만들어 게임을 진행한다.
play메서드는 String으로 결과값을 return 한다.
스트라이크와 볼이 같이 있는 경우 볼을 먼저쓰고, 스트라이크를 쓴다.

라는 프로그래밍 요구사항을 잘 만족하는 함수네요! 💯 
스트라이크, 볼 갯수를 받아 `string`으로 변환하여 반환하는 로직이 깔끔해보여요.
다만, 더 좋은 네이밍을 제안드리자면 `strike`, `ball`의 변수명이 count(갯수)를 의미를 포함한다면 더 직관적인 변수명이 될 것 같아요!

yoonjaehong26 replied:
```
refactor(index.js):변수 가독성을 위한 변수명 변경

변수 타입과 명확성의 부족으로 다음과 같이 변경
ball=>ballcount
strike=>strikecount

전체적은 변수명 통일을 위하여 다음과 같이 변수명 통일
result=>resultstring
```
다음과 같이 변경하였습니다. 네이밍에 대해 잘 알려주셔서 감사합니다!

Songhyejeong commented on src/uiChanger.js:
```diff
@@ -0,0 +1,12 @@
+export function changeUI(resultString) {
+  document.getElementById("result").style.display = "block";
+  document.getElementById("result").innerHTML = resultString;
+  document.getElementById("game-restart-button").style.display = resultString === "3스트라이크" ? "block" : "none";
```
### 🎲 사용자 경험
<img width="200" height="407" alt="스크린샷 2025-09-13 오후 2 20 09" src="https://github.com/user-attachments/assets/d2b3ca00-9e90-43cc-85d4-5c28527294ae" />

**3스트라이크**와 **재시작** 버튼이 있을 경우, 어떤 의도로 재시작 버튼이 나타났는지 사용자는 한 번에 파악하기 힘들 수 있어보여요. 
프론트엔드에서는 사용자의 경험과 직결되어 있는 만큼 사용자 관점에서 항상 생각해보면 좋아요.
예를 들어, 개발자는 정답 및 게임 재시작이라는 의도를 가지고 있지만 사용자는 게임이 일시 중단되어 재시작 버튼이 나타난건가?라고 생각할 수도 있을 거 같아요!

사용자 경험을 더 좋게 개선해보시겠어요?!

yoonjaehong26 replied:
UX에 대해 부족한 부분을 말씀해 주셔서 감사합니다! 


```Js
if (resultString === '3스트라이크') {
    const gameCompleteMessage = document.createElement('div');
    gameCompleteMessage.className = 'completeMessageDiv';
    gameCompleteMessage.textContent = '축하합니다! 3스트라이크를 맞추셨습니다! 재시작 하겠습니까?';
    document.getElementById("result").appendChild(gameCompleteMessage);
  }
```
위와 같이 코드를 추가하려 하였으나 코드를 짜다보니 새로운 요소를 만드는 것이 유지보수에 유용하지 않다 판단하여  아래와 같이 내부데이터만 바꾸는 방식을 택하였습니다.

```js
    if (strikeCount === 3){
      resultString = "🎉축하합니다!!!🎉 정답을 맞추셨습니다. <br> 재시작 하시겠습니까?";
    }
      return resultString;
  }
```

Songhyejeong commented on src/checkInput.js:
```diff
@@ -0,0 +1,26 @@
+import numToArr from './util.js';
+
+function hasDuplicate(input) {
+  const uniqueNumbers = new Set(input);
+  return uniqueNumbers.size !== input.length;
+}
```
모든 함수를 함수 선언문으로 작성해주셨어요.
`함수 선언문`과, `함수 표현식`의 차이는 무엇일까요? 🤔
(호이스팅에 대한 키워드와 함께 공부해보시면 좋아요!)

yoonjaehong26 replied:
제가 몰랐던 새로운 개념을 알려주셔서 감사합니다!!!
***
함수선언문
```js 
function hasDuplicate(input){ ~}
```

함수 표현식
```js 
let hasDuplicate =function(input){ ~}

let hasDuplicate =()=>{ ~}

```
제가 c++함수에만 익숙해서 함수 선언식만 알고 애로우 펑션으로 나타낼 수 있다 정도로 알고 있었는데 
함수 선언식과 함수 표현식 2가지로 함수를 나타낼 수 있고 함수 
표현식을 애로우 펑션으로 축약할 수 있다고 알게 되었습니다.
***
또한 함수 선언문은 호이스팅(코드 실행 전 소속 스코프에 전역으로 선언됨)
으로 인하여 예기치 못한 오류가 발생 할 수 있다고 하네요!


대형프로젝트에서는 메모리 효율이 더 상승한다고 합니다.
***
다만 함수 선언식은 함수 선언 전 호출이 되지 않아 재사용이 많이되는 함수에는 적절하지 않으며 함수이름을 사용하지 않으면 유지보수에 어려움이 있을 수 있다고 합니다. 다만 클라스 내부에서는 호이스팅 문제를 신경쓰지 않아도 되는 것 같습니다.

결론적으로 제가 작성한 함수들은 클라스 내부에 있어 호이스팅을 
신경쓰지 않아도 되나 
필요에 따라 this상속문제에 있어 애로우 펑션 사용이 권장될 것 같습니다.
this상속에 관해서는 아래 질문에서 다루겠습니다.
좋은 질문 주셔서 감사합니다!

Songhyejeong commented on src/checkInput.js:
```diff
@@ -0,0 +1,26 @@
+  return true;
+}
+
+export default function isValidInput(input, digitCount) {
```
이 파일은 모듈화가 깔끔하게 잘 되어 있네요!
입력값을 받아 각 함수에서 단계별로 유효성을 검사한 뒤 최종적으로 isValidInput을 반환하는 흐름이 명확해서 가독성이 좋아보여요!

질문이 하나 있는데, 어떤 모듈에서는 export로, 해당 모듈에서는`export default`로 
각각 다른 방식으로 모듈을 내보내고 있어요!
`export default`로 내보내는 것과, `export`로 내보내는 것의 차이를 설명해주실 수 있나요?

yoonjaehong26 replied:
코드를 짜면서 import문법이 생각이 안나서 검색해 보니 하나를 내보낼때는 export default, 여러개를 내보낼 때에는export를 사용한다고 나와있어 
isValidInput를 하나로 내보내야지! 에만 집중하여 왜 export default를 사용하는 지 생각해 보지 못했습니다.
***
검색해 보니 파일의 대표성을 나타내고 임의의 이름으로 가져 올 수 있어 편리하다고 나와있습니다.
좋은 질문 주셔서 감사합니다!
***
지금의 방식보다 export로 통일하는것이 코드가 더 깔끔할 지 궁금합니다!

Songhyejeong replied:
`export`와 `export default`의 차이를 잘 설명해주셨어요!
`export default`를 사용하면 재홍님께서 정리해주신 것처럼 이름을 자유롭게 변경할 수 있어 여러 불편한 점이 있습니다.
- 이름이 의도치않게 변경될 경우 나중에 혼란스러울 수 있겠죠. 

그럼에도 불구하고 `export default`를 사용하는 경우가 흔한데, 에어비엔비 컨벤션에서도 모듈이 하나의 함수만 내보낼 경우에는 export default를 쓰는 것을 권장합니다.

[에어비엔비 컨벤션 10.6](https://github.com/airbnb/javascript?tab=readme-ov-file#modules--prefer-default-export)
이유는 다음과 같이 나와있습니다.

> Why? To encourage more files that only ever export one thing, which is better for readability and maintainability.

하나의 파일에 하나의 함수만을 내보낸다면, export defualt 명시하는 것이 유지보수 측면에서도 좋고, 읽기에도 편하겠죠?!
재홍님의 코드 같은 경우에 `isValidInput` 함수만을 내보내고 있기 때문에 export defualt를 사용하는 것을 권장드려요!

Songhyejeong commented on src/index.js:
```diff
@@ -0,0 +1,83 @@
+  }
+
+  makeRandomNumber() {
+    const numbers = [];
```
이번 미션에서 제공해드렸던 에어비엔비 컨벤션에는 `var`로 변수를 선언하는 것을 지양하라고 나와있는데요!
`var`를 사용하였을 때 어떤 문제점들이 있을까요?

<details><summary>참고❗</summary>
<p>

> let, const는 JS의 ES6에서 var를 사용했을 때의 문제점을 해결하기 위해 등장했어요.

</p>
</details>

하나 더 질문드리자면,
`let` 그리고 `const`의 차이는 무엇일까요?

yoonjaehong26 replied:
var의 문제점
1.쓰기 애매한 스코프 범위
블록스코프를 쓰는 let,const와 달리 함수단위 스코프를 사용하기 때문에
for문이나 if문을 벗어나 예상치 못한 오류를 일으킬 수 있습니다.
***
2.호이스팅 이슈
위 질문과 같은 원리로 스코프 최상단에 미리 선언되어 예상치 못한 오류를 일으킬 수 있습니다.
***
3.중복 변수선언 가능성
let,const와 달리 중복선언이 되어 같은 개발자가 실수로 같은 변수명으로 선언해도 오류가 나지 않아  변수관리에 어려움이 있습니다.
***
이러한 문제로 쓰게 된 let과 const는 재선언이 불가능하지만
let은 데이터의 값을 바꿀 수 있는 재할당이 가능하지만 const는 불가능 합니다.
따라서 const는 바뀌지 않을 값에 선언하며 let은 유동적인 변수로 사용하게 됩니다.

Songhyejeong replied:
잘 정리해주셨네요 💯 
추가로, 재홍님의 해당 코드에서

```js
  makeRandomNumber() {
    const numbers = [];
    while (numbers.length < this.digitCount) {
      const number = MissionUtils.Random.pickNumberInRange(1, 9);
      if (!numbers.includes(number)) {
        numbers.push(number);
      }
    }
    return numbers;
  }
```

오늘 스터디 시간에 공유드린 내용을 바탕으로
배열인 `numbers`를 `const`로 선언해도 동작하는 이유를 설명해주실 수 있나요?

yoonjaehong26 replied:
넵 위에서 말씀드렸다시피 const는 재할당을 막는 기능을 가지고 있습니다.
***
하지만 const는 원시타입 데이터의 데이터 자체를 저장하여 재할당을 막을 수 있지만 
객체나 배열과 같은 복학타입 데이터의 메모리 주소를 저장하기 때문에 메모리 주소를 바꾸지 않는 내부요소 변경은 가능합니다.
***
위의 변수 `numbers`은 배열타입이므로 .push를 통해 내부요소만 변경하므로 const사용이 가능합니다.

Songhyejeong commented on src/index.js:
```diff
@@ -0,0 +1,83 @@
+
+
+export default class BaseballGame {
+  constructor(digitCount = 3) {
```
digitCount = 3로 넘겨준 이유를 들어볼 수있을까요?
this.digitCount = 3처럼 고정 값으로 인스턴스를 초기화 한다면 어떨까요?

yoonjaehong26 replied:
이 부분은 제가 생각한 확장성을 고려하여 생성자의 인자로 넘겨준 기능입니다.
***
어떻게 하면 코드의 확장성을 고려할 수 있을까를 고민해본 결과 저는 숫자야구게임이 
회원가입 비밀번호 생성 제한 기능이나 검색 기능과 비슷하다고 생각하였습니다.

이에 따라 각 기능을 최대한 함수로 분리하여 특히 checkInput.js의 독립을 통하여 이후 유지보수의 가능성을 생각하였고 추가적으로 자리수변동에 대비하여 하드코딩을 하지 않았습니다.
***

```js
new BaseballGame(4);
//4자리수 야구게임
```
위와 확장성을 고려하여 위와같은 방식을 택하였으나 다시 한번 생각해 보니 
항후 개발 시 더 편리한지는 불명확한 것 같습니다.
이부분에 대해서는 더 생각해 보겠습니다!

Songhyejeong replied:
확장성을 고려하여 작성한 코드였군요! 💯 
확장성을 고려한 것 같다고 예측은 했지만 의도를 듣고 싶었답니다!

Songhyejeong replied:
확장성을 고려하신다면 **매직넘버**라는 키워드도 학습해보시는 것을 추천드립니다!
4와 같은 숫자는 개발자가 추후에 의도를 명확하게 파악하기 힘들 수 있답니다.

Songhyejeong commented on src/index.js:
```diff
@@ -0,0 +1,83 @@
+
+    for (let i = 0; i < computerInputNumbers.length; i++) {
+      if (computerInputNumbers[i] === userInputNumbers[i]) {
+        strike++;
```
이 부분도 airbnb 코드 컨벤션에 나와 있는 내용인데, 증감 연산자로 `++ `대신 `+=`를 사용하는 것을 권장드려요!

yoonjaehong26 replied:
refactor(index.js): airbnb 코드 컨벤션에 따른 연산자 표현 변경

수정 완료하였습니다. 
코드 꼼꼼히 읽어주셔서 감사합니다!!

Songhyejeong commented on src/index.js:
```diff
@@ -0,0 +1,83 @@
+  }
+
+  gameLogic(e) {
+    e.preventDefault();
```
`e.preventDefault()`를 적용해주셨군요 !
preventDefault라는 네이밍을 보니, default로 이벤트 발생시 일어나는 일을 막는 함수 같은데, 
해당 함수 내의 e.preventDefault의 역할은 무엇인가요?

yoonjaehong26 replied:
좋은 질문 감사드립니다!
원래 e.preventdefault가 새로고침을 막는 기능이라고 알고 있었는데
복붙만 하다보니 정작 함수 이름을 재대로 생각해 보지 않았습니다!
***
이번에 알아보니 자동으로 진행되는 기본본동작을 막는 기능으로,
e.preventdefault을 통하여 form,input,<a>작동 시 브라우저가 자동으로 페이지를 이동시키는 것이 기본동작이고 context 우클릭, keydown스크롤,
dragstart 드래그,drop새파일열기등이 기본동작으로,
유저의 의도치 않은 조작을 조절할 수 있는 역활을 수행합니다.

Songhyejeong commented on src/uiChanger.js:
```diff
@@ -0,0 +1,12 @@
+
+export function resetUI() {
+  document.getElementById("user-input").value = "";
+  document.getElementById("result").innerHTML = "";
+  document.getElementById("result").style.display = "none";
```
**resultUI, changeUI 네이밍에 관하여**

네이밍(변수명, 함수명)은 길어도 괜찮습니다. 단순하고, 짧은 네이밍보다, 직관적이면서 긴 네이밍이 더 좋은 네이밍이에요.
그 관점에서 change, reset이라는 네이밍은 의도를 파악하기 쉽지만 뒤에 "UI" 라는 네이밍은 UI를 바꾼다? 어떤 UI를 바꾸는거지?라는 생각이 들었어요.

이해하기 쉬운 네이밍으로 변경하려면 어떤 네이밍이 적합할까요?

yoonjaehong26 replied:
refactor(index.js,uiChanger.js):함수목적성을 강조하기 위한 함수명 변경

resultUI=>resultGameResultUI
changeUI =>changeGameResultUI 
변경완료하였습니다.
좋은 제안 감사드립니다!

Songhyejeong commented on src/checkInput.js:
```diff
@@ -0,0 +1,26 @@
+
+function isEachNumberInRange(input) {
+  for (const char of input) {
+    const number = parseInt(char, 10);
+    if (isNaN(number) || number < 1 || number > 9) {
+      return false;
+    }
```
문자나, 0과 같은 예외까지 꼼꼼하게 처리해주셨네요! 👍 
한 가지 제안을 드리자면,
사용자의 예상치 못한 입력은 생각보다 더 다양할 수 있어요.
3e2나, 2.5처럼요! 이부분을 고려해서 리팩토링을 해보시면 좋을 것 같아요!

yoonjaehong26 replied:
```js
  uiEventListeners() {
    document.getElementById("submit").addEventListener("oninput", this.gameLogic);
    document.getElementById("game-restart-button").addEventListener("oninput", this.restartLogic);
  }
```
원래는 위와 같이 click이 아닌 oninput으로 엄격하게 검사하는 것이 가장 확실하다고 생각하였으나
아래 질문주신 dom 탐색 소요같은 문제가 있을 것 같아 구현하지 않았습니다.
***
검색해 보니 정규표현식을 사용할 수도 있는 것 같아 더 엄격한 제한을 적용해 보겠습니다!
feat(checkInput.js):입력 유효성 검사 정규표현식 활용 조건 추가
```js
if (!input.test(/^ [1-9]+ $/)) {
    return false;
  }
```

Songhyejeong replied:
`oninput`이라는 속성은 처음 알았네요! input값이 변경될 때마다 이벤트가 발생하는 건가요?

제가 이해한 바로는 
```js
document.getElementById("submit").addEventListener("oninput", this.gameLogic);
```
여기서 `getElementById("submit")`는 이벤트 리스너를 등록하는 시점에만 실행됩니다!
이벤트가 발생할 때마다 다시 DOM 탐색을 하는 것으로 헷갈릴 수 있을 것 같아요!
즉, 등록이 끝나면 찾은 요소에 이벤트 핸들러가 연결돼서, 브라우저는 그 핸들러만 호출해주게 되는 것입니당

yoonjaehong26 replied:
네 oninput은 <input>, <select>, <textarea> 태그 내 value값이 변경될 때 마다 함수가 실행됩니다. 
답변해 주신 내용을 찾아보니 '함수가 실행되는 것'과 'dom을 탐색하는 것'에으로 나누어 생각하지 못하고 그저 해당 라인의 코드가 실행된다고 생각하여  소요가 발생한다고 생각했던 것 같습니다.

js가 쉬운 언어라고 생각했는데 말씀해 주신대로 공부할 것이 많은 언어인 것 같습니다,,,

Songhyejeong commented on src/checkInput.js:
```diff
@@ -0,0 +1,26 @@
+import numToArr from './util.js';
```
해당 import문은 사용하지 않고 있어서 지워주셔도 좋을 것 같아요!😶

yoonjaehong26 replied:
refactor(checkInput.js):불필요한 numToArr함수 import삭제

넵 코드 정리하기 전에는 필요했었는데 리펙터링 이후 없애는 것이 더 깔끔한 것 같습니다!

Songhyejeong commented on src/index.js:
```diff
@@ -0,0 +1,83 @@
+    const userInput = document.getElementById("user-input").value;
+    if (!isValidInput(userInput, this.digitCount)) {
+      alert("잘못된 입력입니다. 중복되지 않는 서로 다른 3개의 숫자를 입력하세요");
+      document.getElementById("user-input").value = "";
```
`document.getElementById(user-input)` 호출시에 어떤 일이 일어날까요?

<details><summary>고민 후 토글을 열어주세요❗</summary>
<p>

`document.getElementById(user-input)` 호출시에 브라우저가 DOM 트리를 탐색해서 user-input에 해당하는 HTMLElement 객체를 찾아 반환한답니다.
현재 재홍님의 코드에서 `document.getElementById("user-input").value` 로 같은 요소를 DOM에서 두 번 찾고 있어요.

현재는 DOM을 2번만 탐색하지만, `document.getElementById("user-input").value;` 코드를 사용하는 곳이 많아진다면 DOM에서 요소를 계속 찾아야하는 비효율성이 생길 거예요. 

</p>
</details> 


더 효율적으로 동작하기 위해서는 어떻게 로직을 수정하면 좋을까요?!

yoonjaehong26 replied:
.addEventListener가 메모리 누수때문에 사용을 최대한 적게 해야 한다는 것을 알고 있었는데 document.getElementById와 같은 dom 탐색시에도 리소스 낭비가 발생하는군요! 
***
addEventListener를 하나의 변수로 저장해 두고 다른 코드에서 
사용하듯이 document.getElementById또한 같은 방법으로 변수에 저장하면 
될 것 같습니다.
***
개인적으로 다른사람의 코드를 읽을 때 document.getElementById를 변수로 저장하고 사용하는코드에서 한눈에 보기 어려운 경우가 많아 저는 그냥 document.getElementById로 작성하는 것이 가독성이 높다고 생각하였는데
말씀해 주신 부분을 이해하고 나니 왜 변수로 저장하였고
검색해보니 $변수명 이나 변수명Element로 작성하는 관행이 있어 이러한 부분에 
대하여 제 지식이 부족했던 것 같습니다.
***
앞으로 document.getElementById 사용 시 중복되는 부분만 리소스 관리를 
위해 변수로 저장하는 것이 나은지,
document.getElementById 전부를 변수로 저장하여 가독성과 통일성을 높이는 것이 나은 지 여쭤 보고 싶습니다!

yoonjaehong26 replied:
`getElementById()`
`getElementByClassName()`
`querySelector()`
`querySelectorAll()`
검색해 보니 아래로 내려갈수록 가독성과 범위가 늘어나는 장점이 있지만 dom탐색시간이 늘어나고 동적인 dom추가를 탐지하지 못한다고 하는데 보통 어느 요소탐색문법을 사용하는 게 좋을 지 여쭤보고 싶습니다!

Songhyejeong replied:
요소탐색문법에 대해서 고민을 해주셨군요.
저도 숫자 야구 게임 미션을 할 때 getElementById(), querySelector()의 차이에 대해서 궁금했었답니다!

```js
getElementById(),
getElementsByClassName(),
querySelector(),
querySelectorAll()
```
를 각각 보면서 비교해보도록할게요!

1. `getElementsByClassName() `vs.  `getElementById()`
`getElementsByClassName()` 먼저 보자면 이 아이는 `className`에 해당하는 여러 엘리먼트들을 가지고 올 수 있습니다. 
HTML에서 공통된 CSS 속성을 가진 아이들을 class로 묶는데 해당 class를 가진 요소들을 가지고 옵니다. 반면,  `getElementById() ` 같은 경우에는 **id는 고유한 값**입니다. 따라서 하나의 요소를 찾는데에 보통 사용됩니다.

2. `querySelector()` vs. `querySelectorAll()`
`querySelector()`는 우선 CSS 선택자 문법을 사용해서 요소를 탐색합니다.
재홍님께서 말씀해주신 것 처럼 `querySelectorAll()`는 이름에서 알 수 있듯이 모든 요소를 탐색합니다. 당연히 무겁게 되겠지요.
우선, 저희는 여러개의 요소를 탐색하는 것이 아닌 돔에서 하나의 요소만을 찾고 있습니다.
따라서 여러 요소를 탐색하고 무거운 `querySelectorAll()`,`getElementsByClassName() `
는 제외하고 

3. **getElementById()** 그리고 **querySelector()** 에 대해 비교해보도록하겠습니다.
우선 **getElementById()** 는 가볍습니다. 좁고 명확하다는 장점도 있습니다.  
하지만 querySelector는 CSS 선택자를 이용할 수 있어 더 복잡한 조건을 만족하는 요소를 선택할 수 있습니다. 

```js
document.querySelector('.form-container input[type="password"]');

document.querySelector('.menu > li:first-child');
```

따라서 지금처럼 id로 간단하게 탐색할 수 있을 때는 `getElementsByClassName()`로
중첩 선택 or 자식 선택 처럼 복잡한 예시는 `querySelector`로 탐색할 수 있을 것 같아요!

yoonjaehong26 replied:
답변 감사합니다! 이번 과제는 id를 이용하고 dom트리구조가 복잡하지 않아` getElementById()`으로 사용 하지만 
이후 과제에서 여러 dom요소를 다루게 된다면 경우에 따라 잘 선택해 보겠습니다!

Songhyejeong commented on src/index.js:
```diff
@@ -0,0 +1,83 @@
+
+  uiEventListeners() {
+    document.getElementById("submit").addEventListener("click", this.gameLogic.bind(this));
+    document.getElementById("game-restart-button").addEventListener("click", this.restartLogic.bind(this));
```
이렇게 하면 `bind(this)`를 사용해 `this.gameLogic`의 this를 BaseballGame 인스턴스로 고정할 수 있겠군요! 👍 

아래처럼 화살표 함수를 사용해도 동일한 효과를 얻을 수 있을 것 같은데, 두 방법은 어떤 차이가 있을까요?
```js
document.getElementById("submit")
  .addEventListener("click", (e) => this.gameLogic(e));
```

yoonjaehong26 replied:
세번째 질문에서 주신 함수 선언식과 함수표현식에 대해 알아보면서 추가로 알게된 this상속에 대하여 정리해 보았습니다. this문법에 대해 아직 햇갈리는 부분이 많아 
공부해나가고 있습니다.
***

함수 표현식,함수 선언식(화살표 함수 제외)의 this:기본적으로 전역범위(window),
객체 내 '호출'시 this: '호출'된 객체를 스코프로 가짐

화살표 함수:자신만의 this가 없으므로 상위 스코프기준으로 범위가 생성
***
따라서 .bind(this)사용 시 '호출'이 기준이 아닌 '선언'이 기준이 되어 함수 재사용이 가능하지만 처음 이 함수만 볼 때 코드를 볼 때 가독성이 떨어져 보일 수 있을 것 같습니다.

반대로 화살표 함수는 함수기준 상위 스코프만 보면 되어서 가독성이 더 증가하는 것 같습니다!

Songhyejeong reviewed (changes requested):
안녕하세요 재홍님! **_@yoonjaehong26_** 👋🏼
저도 첫 미션의 리뷰어로서 만나뵙게 되어 영광입니다!
아쉽게도 리뷰어로서는 자주 뵙지 못할 것 같지만 스터디에서 매주 뵙게 되니 스터디에서 많은 얘기 나누면 좋을 것 같아요.

리뷰에 앞서, 코드에서 재홍님과 나누면 좋을만한 주제를 위주로 질문을 많이 남겨보았어요!

## 🎲 전반적인 숫자 야구 미션 리뷰
이번 첫 미션이 요구사항도 많고, 컨벤션도 지켜야하고 기능 구현에 클린 코드까지 지켜야할 것들이 많았는데 너무 잘해주신 것 같아요! 
커밋이나 PR도 신경써서 작성해주셔서 좋았어요.

>1. 체계적인 코드구성을 위해 기능들을 최대한 함수로 구현과
>4. 복잡하거나 신기능보다는 최대한 기본적인 코드들로 기능 구하여 사이드이펙트 최소화

재홍님께서 신경 써주셨다고 하셨던 것들이 정말 잘 지켜진 것 같았어요.
특히 **모듈을 기능별로 분리하신 점**과, **얼리리턴, 자바스크립트 내장 함수**까지 잘 활용하신 점이 인상 깊었습니다.


> 1.< input>태그에서 입력데이터 프롬프트 인젝션방지
=>초기에는 배열로 변환 후 유효성검사를 하였으나 배열로 변환과정에서 사이드이펙트가 있을 것을 예상하여 함수를 통한 유효성검사처리 후 배열로 변환

코드를 작성할 때 단순히 기능 구현에 초점을 맞추는 것이 아닌, 하나의 코드도 고민해주신 점이 인상 깊네요!

한 가지 궁금한 것이 있는데,
혹시 배열로 변환하는 과정에서 어떤 **사이드 이펙트**가 발생할 수 있는지 설명해주실 수 있나요??

> 3.코드 커밋 단위 구상의 어려움
=>위에서 말씀드린 대로 최대한 함수단위로 프로그렘을 작성하고 싶어 함수를 기준으로 커밋단위를
구상하였으나 함수로 구현되지 않는 부분, (feat(index.js):함수통합 밎 class구현)와 같은 코드구조 변환의 커밋에서 리뷰어가 이해하기 쉬운 단순한 변화가 아니게 되어 이번 과제 중 가장 아쉬운 부분이 되었습니다.

### 커밋 메시지에 관하여
커밋의 단위를 나누려는 노력이 느껴졌답니다 :) 
리뷰를 하다보면 커밋별로 변경 사항을 확인해야하는 일이 생겨요. 그럴 때 저는 주로 커밋 메시지를 통해서 해당 커밋 내역을 확인하는데요!
커밋 메시지에 **“사용자 입력의 길이를 검사하는 로직 구현”** 처럼 상세한 내용을 적어주시면, 리뷰어가 더욱 편하게 이해할 수 있답니다!
함수로 기능을 분리하고 커밋 단위를 쪼개셔도 좋지만, 함수가 작거나 변경사항이 많을 경우 기능이나 하나의 작업에 대한 내용을 포함해주시면 더 좋은 커밋이 될거예요.

### 기능 목록(README)
또한 구현을 하다보면 새로운 요구사항이 보이고 예외 처리해야할 부분도 생기실 거예요.
그럴 때마다 기능 목록를 업데이트 해주시면 방치되는 문서가 아닌 계속해서 업데이트되는 문서로 관리하실 수 있어요.


## ✍🏼 질문에 대한 답변

> 위 첫 질문과 같이 타입스크립트가 아닌 js에서 데이터 타입 핸들링하는 더 나은 방법이 있는지 궁금합니다.

js는 동적 타입 언어라서 타입을 미리 확인하고 런타임 시점 이전에 막을 수는 없습니다. 
대신 코드 상으로는 타입을 핸들링 할 수 있는데요. 
질문을 주신의도가 input의 유효성 검사를 하기 위함이었다면 재홍님께서 작성해주신 것 처럼 `isNaN`으로 판별하여도 되고, `typeof`로 타입을 검사할 수도 있답니다!


> 코드구조나 함수,변수명들이 가독성이 괜찮은지와 개선방안이 궁금합니다.

전반적인 코드 구조, 모듈 분리 및 함수의 가독성은 좋았답니다!
개별 함수, 변수명에 대한 개선 제안을 답변으로 드렸으니 확인해주세요 :)

Commit:
refactor(index.js):변수 가독성을 위한 변수명 변경

변수 타입과 명확성의 부족으로 다음과 같이 변경
ball=>ballcount
strike=>strikecount

전체적은 변수명 통일을 위하여 다음과 같이 변수명 통일
result=>resultstring

Commit:
feat(index.js) 3스트라이크 시 메세지를 수정하여 게임완료를 명확하게 알림

Commit:
refactor(index.js): airbnb 코드 컨벤션에 따른 연산자 표현 변경

Commit:
refactor(index.js,uiChanger.js):함수목적성을 강조하기 위한 함수명 변경

Commit:
refactor(checkInput.js):불필요한 numToArr함수 import삭제

Commit:
refactor(index.js):에어비엔비 코드 컨벤션에 따른 연산자 좌우 빈칸 확보

Commit:
feat(checkInput.js):입력 유효성 검사 정규표현식 활용 조건 추가

Songhyejeong commented on src/checkInput.js:
```diff
@@ -0,0 +1,27 @@
+  }
+  if (!input.test(/^ [1-9]+ $/)) {
+    return false;
+  }
```
정규식으로 인풋의 유효성을 검사하셨군요 ! 👍

Songhyejeong commented on src/index.js:
```diff
@@ -0,0 +1,87 @@
+      resultString += `${strikeCount}스트라이크`;
+    }
+
+    if (strikeCount === 3){
+      resultString = "🎉축하합니다!!!🎉 정답을 맞추셨습니다. <br> 재시작 하시겠습니까?";
+    }
```
사용자가 게임의 결과를 직관적으로 알 수 있게 되었네요 !👍

Songhyejeong commented on src/index.js:
```diff
@@ -0,0 +1,87 @@
+  restartLogic = () => {
+    this.computerInputNumbers = this.makeRandomNumber();
+    resetGameResultUI();
+  }
```
중간 중간 세미콜론이 빠져있어 보여요. ESlint와 같이 포매팅을 자동으로 해주는 도구가 있어요! 
키워드에 대해 공부해보시고, 적용해보시면 좋을 거 같아요! (다음 미션부터 적용하셔도 좋습니다)

Songhyejeong reviewed (approved):
안녕하세요 재홍님! 변경 사항과 답변 모두 확인했습니다.
깊이 있게 공부해주신 덕분에 의미 있는 대화를 나눌 수 있었던 것 같아요.

추가로, 지난번에 말씀드렸던 typeof 같은 수동 타입 검사 방법 외에도 JSDoc이라는 방식이 있더라고요.
자바스크립트처럼 타입을 명시하지 않는 환경에서도 JSDoc 주석을 활용하면 에디터/툴이 **타입 검사를 지원해** 오류를 미리 잡을 수 있습니다.
TypeScript는 무겁다는 단점이 있는데, JSDoc은 비교적 가볍게 타입 검사의 이점을 누릴 수 있다는 장점이 있습니다.

추가로 Tsdoc이라는 것도 있는데, TS 프로젝트에서 문서화를 표준화하기 위한 주석 문법이라고 합니다!
저도 프로젝트에 적용하고 사용하고 있답니다

[JSDoc 링크](https://poiemaweb.com/jsdoc-type-hint)
이 자료 한번 읽어보시면 좋을 것 같아요!

우선 충분히 리뷰 반영해주신 것 같아 approve하겠습니다!

Commit:
refactor(checkInput.js):중복된 숫자검증 제거

Commit:
fix(uiChanger.js):이전 ui 변경관련 커밋에서 반영 안된 부분 수정

이전 5a2de4fc7aa2800a6cbb6a9c4ab63e7dbbfd02cb커밋에서
 index.js getPlayResultString함수에서 바꾼 결과를 changeGameResultUI함수에서 반영이 안된 부분 수정
